bokehjs/src/lib/models/formatters/datetime_tick_formatter.ts

```diff
@@ -234,34 +243,81 @@ export class DatetimeTickFormatter extends TickFormatter {
     return s
   }
 
-  _add_context(tick: number, label: string, i: number, N: number, resolution: ResolutionType): string {
-    const loc = this.context_location
-    const which = this.context_which
+  _compute_context_labels(ticks: number[], resolution: ResolutionType): string[] {
+    const {context} = this
 
-    if (this.context == null) {
-      return label
+    let context_labels: string[]
+    if (isString(context!)) {
+      context_labels = []
+      for (const tick of ticks) {
+        context_labels.push(_strftime(tick, context))
+      }
+    } else {
+      context_labels = context!.doFormat(ticks, {loc: 0}, resolution)
+    }
+
+    const which = this.context_which
+    const N = context_labels.length
+    for (let i=0; i<context_labels.length; i++) {
+      if ((which == "start" && i != 0) ||
+        (which == "end" && i != N-1) ||
+        (which == "center" && i != Math.floor(N/2))) {
+        context_labels[i] = ""
+      }
     }
+    return context_labels
+  }
 
-    if ((which == "start" && i == 0) ||
-        (which == "end" && i == N-1) ||
-        (which == "center" && i == Math.floor(N/2)) ||
-        (which == "all")) {
+  _build_full_labels(base_labels: string[], context_labels: string[]): string[] {
+    const loc = this.context_location
+    const full_labels: string[] = []
 
-      const context = isString(this.context) ?
-        _strftime(tick, this.context) : this.context.doFormat([tick], {loc: 0}, resolution)[0]
+    if (context_labels.every(v => v === "")) {
+      return base_labels
+    }
 
-      if (context == "") {
-        return label
-      }
+    for (let i=0; i<base_labels.length; i++) {
+      const label = base_labels[i]
+      const context = context_labels[i]
+      let full_label: string
 
+      // In case of above and below blank strings are not trimmed in order to
+      // keep the same visual format across all ticks.
       switch (loc) {
-        case "above": return `${context}\n${label}`
-        case "below": return `${label}\n${context}`
-        case "left":  return `${context} ${label}`
-        case "right": return `${label} ${context}`
+        case "above":
+          full_label = `${context}\n${label}`
+          break
+        case "below":
+          full_label = `${label}\n${context}`
+          break
+        case "left":
+          full_label = context == "" ? label : `${context} ${label}`
+          break
+        case "right":
+          full_label = context == "" ? label : `${label} ${context}`
+          break
```

Comment:
To avoid mutation and excessive changes:
```ts
const full_label = (() => {
  switch (loc) {
    case "above": return `${context}\n${label}`
    case "below": return `${label}\n${context}`
    case "left":  return context == "" ? label : `${context} ${label}`
    case "right": return context == "" ? label :`${label} ${context}`
  }
})()
```